File: app/server/notify/channels/ntfy.ts
/**
 * Ntfy 推送渠道
 */
import Logger from '../../utils/logger';
import { NotifyChannel, NotifyResult } from '../types';
import { $ } from '../httpClient';
import { hasConfig, getConfig } from '../config';

const logger = Logger.child({ channel: 'ntfy' });

/**
 * RFC2047 编码
 */
function encodeRFC2047(text: string): string {
    const encodedBase64 = Buffer.from(text).toString('base64');
    return `=?utf-8?B?${encodedBase64}?=`;
}

export const ntfyChannel: NotifyChannel = {
    name: 'ntfy',
    enabled: false,

    async send(text: string, desp: string): Promise<NotifyResult> {
        return new Promise((resolve) => {
            if (!hasConfig('NTFY_TOPIC')) {
                resolve({ success: false, message: 'NTFY_TOPIC 未配置' });
                return;
            }

            const NTFY_URL = getConfig('NTFY_URL') || 'https://ntfy.sh';
            const NTFY_TOPIC = getConfig('NTFY_TOPIC');
            const NTFY_PRIORITY = getConfig('NTFY_PRIORITY') || '3';
            const NTFY_TOKEN = getConfig('NTFY_TOKEN');
            const NTFY_USERNAME = getConfig('NTFY_USERNAME');
            const NTFY_PASSWORD = getConfig('NTFY_PASSWORD');
            const NTFY_ACTIONS = getConfig('NTFY_ACTIONS');

            const url = `${NTFY_URL}/${NTFY_TOPIC}`;
            const headers: Record<string, string> = {
                Title: encodeRFC2047(text),
                Priority: NTFY_PRIORITY,
                Icon: 'https://qn.whyour.cn/logo.png',
            };

            // 认证
            if (NTFY_TOKEN) {
                headers['Authorization'] = `Bearer ${NTFY_TOKEN}`;
            } else if (NTFY_USERNAME && NTFY_PASSWORD) {
                headers['Authorization'] = `Basic ${Buffer.from(`${NTFY_USERNAME}:${NTFY_PASSWORD}`).toString('base64')}`;
            }

            // Actions
            if (NTFY_ACTIONS) {
                headers['Actions'] = encodeRFC2047(NTFY_ACTIONS);
            }

            $.post({ url, body: desp, headers }, (err, _resp, data) => {
                if (err) {
                    logger.error({ err }, 'Ntfy 发送失败');
                    resolve({ success: false, error: err as Error });
                } else {
                    const result = data as { id?: string };
                    if (result?.id) {
                        logger.info('Ntfy 发送成功');
                        resolve({ success: true, message: '发送成功' });
                    } else {
                        logger.warn({ body: data }, 'Ntfy 发送失败');
                        resolve({ success: false, message: '发送失败' });
                    }
                }
            });
        });
    }
};

if (hasConfig('NTFY_TOPIC')) {
    ntfyChannel.enabled = true;
}

export default ntfyChannel;

File: app/server/internal/notify/channels/ntfy.go
package channels

import (
	"encoding/base64"
	"log/slog"

	"github.com/sushazhi/fnos-logmanager/internal/notify"
)

func encodeRFC2047(text string) string {
	encoded := base64.StdEncoding.EncodeToString([]byte(text))
	return "=?utf-8?B?" + encoded + "?="
}

type Ntfy struct {
	enabled bool
}

func (c *Ntfy) Name() string {
	return "ntfy"
}

func (c *Ntfy) Enabled() bool {
	return c.enabled
}

func (c *Ntfy) SetEnabled(enabled bool) {
	c.enabled = enabled
}

func (c *Ntfy) Send(text, desp string) notify.NotifyResult {
	if !notify.HasConfig("NTFY_TOPIC") {
		return notify.NotifyResult{Success: false, Message: "NTFY_TOPIC 未配置"}
	}

	ntfyURL := notify.GetConfig("NTFY_URL")
	if ntfyURL == "" {
		ntfyURL = "https://ntfy.sh"
	}
	ntfyTopic := notify.GetConfig("NTFY_TOPIC")
	ntfyPriority := notify.GetConfig("NTFY_PRIORITY")
	if ntfyPriority == "" {
		ntfyPriority = "3"
	}
	ntfyToken := notify.GetConfig("NTFY_TOKEN")
	ntfyUsername := notify.GetConfig("NTFY_USERNAME")
	ntfyPassword := notify.GetConfig("NTFY_PASSWORD")
	ntfyActions := notify.GetConfig("NTFY_ACTIONS")

	url := ntfyURL + "/" + ntfyTopic

	headers := map[string]string{
		"Title":    encodeRFC2047(text),
		"Priority": ntfyPriority,
		"Icon":     "https://qn.whyour.cn/logo.png",
	}

	// Authentication
	if ntfyToken != "" {
		headers["Authorization"] = "Bearer " + ntfyToken
	} else if ntfyUsername != "" && ntfyPassword != "" {
		auth := base64.StdEncoding.EncodeToString([]byte(ntfyUsername + ":" + ntfyPassword))
		headers["Authorization"] = "Basic " + auth
	}

	// Actions
	if ntfyActions != "" {
		headers["Actions"] = encodeRFC2047(ntfyActions)
	}

	resp, err := notify.HTTPPost(url, notify.HttpRequestOptions{
		Body:    desp,
		Headers: headers,
	})
	if err != nil {
		slog.Error("Ntfy 发送失败", "error", err)
		return notify.NotifyResult{Success: false, Message: err.Error(), Err: err}
	}

	if resp.StatusCode == 200 {
		slog.Info("Ntfy 发送成功")
		return notify.NotifyResult{Success: true, Message: "发送成功"}
	}

	slog.Warn("Ntfy 发送失败", "status", resp.StatusCode, "body", resp.Body)
	return notify.NotifyResult{Success: false, Message: "发送失败"}
}

func init() {
	ch := &Ntfy{}
	if notify.HasConfig("NTFY_TOPIC") {
		ch.SetEnabled(true)
	}
	notify.Registry.Register(ch)
}
